src/spatial.rs

```diff
@@ -74,6 +74,46 @@ pub fn expand_for_over(old_dims: &[usize]) -> Vec<(Vec<usize>, Vec<usize>)> {
         .collect()
 }
 
+pub fn expand_for_up(old_dims: &[usize], position: usize) -> Vec<(Vec<usize>, Vec<usize>)> {
+    // Handle edge case for empty dims - return [2]
+    if old_dims.is_empty() {
+        let new_dims = vec![2];
+        let reorganization_pattern = vec![0];
+        return vec![(new_dims, reorganization_pattern)];
+    }
+    
+    // Always apply up (append 2) followed by over transformations
+    let up_dims = create_up_dims(old_dims);
```

Comment:
Do not chain up and over. Look at the tests. The tests show that over can take a base shape (all 2s) and do over transforms without incident. That output should be mixed in with a single up result which is what happens when you find the next up shape, pad, and apply reorganization.